Tests/RegexTests/MatchTests.swift

```diff
@@ -0,0 +1,467 @@
+    matchTest(
+      #"[ab[:space:]\d[:^upper:]cd]"#,
+      input: "123abcxyz", match: "1", xfail: true)
+    matchTest(
+      #"[ab[:space:]\d[:^upper:]cd]"#,
+      input: "xyzabc123", match: "x", xfail: true)
+    matchTest(
+      #"[ab[:space:]\d[:^upper:]cd]"#,
+      input: "XYZabc123", match: "a", xfail: true)
+    matchTest(
+      #"[ab[:space:]\d[:^upper:]cd]"#,
+      input: "XYZ abc123", match: " ", xfail: true)
+
+    matchTest("[[[:space:]]]", input: "123 abc xyz", match: " ", xfail: true)
+
+    matchTest("[[:alnum:]]", input: "[[:alnum:]]", match: "a", xfail: true)
+    matchTest("[[:blank:]]", input: "123\tabc xyz", match: "\t", xfail: true)
+    matchTest(
+      "[[:graph:]]",
+      input: "\u{7}\u{1b}\u{a}\n\r\t abc", match: "a", xfail: true)
+    matchTest(
+      "[[:print:]]",
+      input: "\u{7}\u{1b}\u{a}\n\r\t abc", match: " ", xfail: true)
+    matchTest(
+      "[[:word:]]",
+      input: "\u{7}\u{1b}\u{a}\n\r\t abc", match: "a", xfail: true)
+    matchTest(
+      "[[:xdigit:]]",
+      input: "\u{7}\u{1b}\u{a}\n\r\t abc", match: "a", xfail: true)
+
+    matchTest("[[:isALNUM:]]", input: "[[:alnum:]]", match: "a", xfail: true)
+    matchTest("[[:AL_NUM:]]", input: "[[:alnum:]]", match: "a", xfail: true)
+    matchTest("[[:script=Greek:]]", input: "123αβγxyz", match: "α", xfail: true)
+
+    // MARK: Operators
+
+    matchTest(
+      #"[a[bc]de&&[^bc]\d]+"#, input: "123bcdxyz", match: "d")
+
+    // Empty intersection never matches, should this be a compile time error?
+    // matchTest("[a&&b]", input: "123abcxyz", match: "")
+
+    matchTest(
+      "[abc--def]", input: "123abcxyz", match: "a")
+
+    // We left-associate for chained operators.
+    matchTest(
+      "[ab&&b~~cd]", input: "123abcxyz", match: "b")
+    matchTest(
+      "[ab&&b~~cd]", input: "123acdxyz", match: "c") // this doesn't match NSRegularExpression's behavior
+
+    // Operators are only valid in custom character classes.
+    matchTest(
+      "a&&b", input: "123a&&bcxyz", match: "a&&b")
+    matchTest(
+      "&?", input: "123a&&bcxyz", match: "")
+    matchTest(
+      "&&?", input: "123a&&bcxyz", match: "&&")
+    matchTest(
+      "--+", input: "123---xyz", match: "---")
+    matchTest(
+      "~~*", input: "123~~~xyz", match: "~~~")
+
+    // MARK: Quotes
+
+    matchTest(
+      #"a\Q .\Eb"#,
+      input: "123a .bxyz", match: "a .b", xfail: true)
+    matchTest(
+      #"a\Q \Q \\.\Eb"#,
+      input: #"123a \Q \\.bxyz"#, match: #"a \Q \\.b"#, xfail: true)
+    matchTest(
+      #"\d\Q...\E"#,
+      input: "Countdown: 3... 2... 1...", match: "3...", xfail: true)
+
+    // MARK: Comments
+
+    matchTest(
+      #"a(?#comment)b"#, input: "123abcxyz", match: "ab")
+    matchTest(
+      #"a(?#. comment)b"#, input: "123abcxyz", match: "ab")
+
+    // MARK: Quantification
+
+    matchTest(
+      #"a{1,2}"#, input: "123aaaxyz", match: "aa", xfail: true)
+    matchTest(
+      #"a{,2}"#, input: "123aaaxyz", match: "", xfail: true)
```

Comment:
Why is the match empty? This would be eager by default, so I would expect the first "aa" to match.

Reply:
(note that PCRE doesn't parse the `{,2}` as a range, but Python, Oniguruma, and others do)

Reply:
Ugh, never mind, it successfully matches the empty range at the start.